app/code/Magento/Ui/view/base/web/js/grid/masonry.js

```diff
@@ -0,0 +1,280 @@
+                            running = false;
+                        });
+                    } else {
+                        setTimeout(function () {
+                            handler();
+                            running = false;
+                        }, 66);
+                    }
+                }
+            });
+        },
+
+        /**
+         * Set layout styles inside the container
+         */
+        setLayoutStyles: function () {
+            var containerWidth = parseInt(this.container.clientWidth, 10),
+                rowImages = [],
+                ratio = 0,
+                rowHeight = 0,
+                calcHeight = 0,
+                isLastRow = false,
+                rowNumber = 1;
+
+            this.setMinRatio();
+
+            this.rows().forEach(function (image, index) {
+                ratio += parseFloat((image.width / image.height).toFixed(2));
+                rowImages.push(image);
+
+                if (ratio < this.minRatio && index + 1 !== this.rows().length) {
+                    // Row has more space for images and the image is not the last one - proceed to the next iteration
+                    return;
+                }
+
+                ratio = Math.max(ratio, this.minRatio);
+                calcHeight = (containerWidth - this.imageMargin * rowImages.length) / ratio;
+                rowHeight = calcHeight < this.maxImageHeight ? calcHeight : this.maxImageHeight;
+                isLastRow = index + 1 === this.rows().length;
+
+                this.assignImagesToRow(rowImages, rowNumber, rowHeight, isLastRow);
+
+                rowImages = [];
+                ratio = 0;
+                rowNumber++;
+
+            }.bind(this));
+        },
+
+        /**
+         * Apply styles, css classes and add properties for images in the row
+         *
+         * @param {Object[]} images
+         * @param {Number} rowNumber
+         * @param {Number} rowHeight
+         * @param {Boolean} isLastRow
+         */
+        assignImagesToRow: function (images, rowNumber, rowHeight, isLastRow) {
+            var imageWidth;
+
+            images.forEach(function (img) {
+                imageWidth = rowHeight * (img.width / img.height).toFixed(2);
+                this.setImageStyles(img, imageWidth, rowHeight);
+                this.setImageClass(img, {
+                    bottom: isLastRow
+                });
+                img.rowNumber = rowNumber;
+            }.bind(this));
+
+            images[0].firstInRow = true;
+            images[images.length - 1].lastInRow = true;
+        },
+
+        /**
+         * Wait for container to initialize
+         */
+        waitForContainer: function (callback) {
+            if (typeof this.container === 'undefined') {
+                setTimeout(function () {
+                    this.waitForContainer(callback);
+                }.bind(this), 500);
+            } else {
+                callback();
```

Comment:
Is it possible for `this.container` to be truthy on the first invocation? If so, this method [releases zalgo](https://blog.izs.me/2013/08/designing-apis-for-asynchrony).

We want to avoid methods that are sometimes sync and sometimes async, because they're hard to reason about.

The solution would be to wrap the `callback();` inside the `else` with `setTimeout(callback, 0)` to make sure it's not invoked until the next tick of the event loop

Reply:
Done